mantidproject/mantid
# Logbook generator for ILL data

**Description of work.**

Generates logbook (`TableWorkspace`) containing all default logbook entries and headers that are pre-defined in the IPF relevant to the chosen instrument. Currently supported instruments: D11, D11B, D16, D22, D22B, D33, IN4, IN5, IN6, PANTHER, D7 (list of parameters not finalized) form the ILL.

There are three levels of the entries: default, optional and custom. Default entries are always read and included in the logbook. Optional entries are pre-defined in the relevant IPF and are included in the logbook if a header for the entry is specified. The last one is added when the full path in the NeXus is provided as a comma separated list.

Basic binary arithmetic operations: addition (the only one supported if the entry data is a string), subtraction, multiplication and division on entries data are supported. At the moment, no support is offered for organizing the operations using parentheses or scaling by read data by constant numbers.

**To test:**

1. From the workbench, specify ILL/D7 UnitTest data directory as 'Directory', and a numor range "394458, 394882". Giving no numor range will result in some rows being filled with "Not found" messages, as this directory contains also processed data that has different structure than raw data. Setting 'D7' as the instrument is mandatory.
2. Inspect the created `TableWorkspace`:
a) How many rows does it have? It should be exactly the number of files contained in the directory, unless the numor range has been speciifed and excludes these files. In the provided example range, there should be 2 rows.
b) What data is contained in the table? If the OptionalHeaders were not specified, it should contain only the entries specified in the IPF, currently: 'run_number', 'wavelength',  'experiment_identifier', 'start_time', 'end_time', 'duration'. Do the values make sense?
3. Specify a path and a name for the `OutputFile` property. Check if the created .csv file can be imported by your preferred spreadsheet software, e.g. LibreOffice Calc or Microsoft Excel.
3. Try setting a different instrument:
a) One that does not belong to the ILL should not go through validation, 
b) One that is not currently supported should raise an error that no default parameters are specified,
c) One that is supported but is not D7: then the output should contain many fields filled with "Not found" message, for every entry that was not found in the NeXus files. 



Fixes #30244 . 


Release notes added.



---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
